crates/server/src/server/server_future.rs

```diff
@@ -1214,6 +1259,10 @@ fn sanitize_src_address(src: SocketAddr) -> Result<(), String> {
     }
 }
 
+fn is_unrecoverable_socket_error(err: &io::Error) -> bool {
+    matches!(err.raw_os_error(), Some(libc::EBADF | libc::ENOTCONN))
```

Comment:
It would be nice to avoid pulling in libc for this -- for `ENOTCONN` I think we could use `err.kind()` matching `io::ErrorKind::NotConnected`?

Not sure where you're getting `EBADF` from.

Reply:
`EBADF` is what I get from `TcpListener::accept` after calling `pid_shutdown_sockets`. Subsequent calls return `ECONNABORTED` (`ErrorKind::ConnectionAborted`). I suppose we could match `ErrorKind::NotConnected | ErrorKind::ConnectionAborted`?

Reply:
If that works, sounds good!

Reply:
TCP seems a bit hopeless. With `TcpListener` from `std`, `accept` returns immediately when the socket dies, but tokio's equivalent just never returns. I suppose this is still an improvement to prevent the pointless retries.

Reply:
Might be worth asking some questions in the Tokio issue tracker or on Discord about the behavior there? Sounds suspect.

Reply:
I've opened an issue there. How do you think we should handle this in the meantime?

Reply:
If the Tokio listener just yields pending forever, that seems okay to me for now.